keps/sig-network/3705-cloud-node-ips/README.md

```diff
@@ -0,0 +1,638 @@
+  the needed algorithms for the new behaviors.
+
+### Non-Goals
+
+- Changing the behavior of nodes using legacy cloud providers.
+
+- Changing the node-IP-selection behavior in any currently-existing
+  Kubernetes cluster. This means that the default behavior when
+  `--node-ip` is not specified will remain the same, and the behavior
+  of any currently-allowed `--node-ip` value will remain the same. New
+  behavior will only be triggered by `--node-ip` values that would
+  have been rejected by kubelet in older clusters.
+
+- Adding the ability for nodes in clusters with cloud providers to use
+  node IPs that are not already known to the cloud provider. (In
+  particular, this implies that we will continue to not support
+  dual-stack nodes in clouds that do not themselves support
+  dual-stack.)
+
+- Improving the behavior of autodetecting the node IP in any other
+  ways. (Eg, being able to pass a CIDR rather than an IP.) There was
+  some discussion of ways we might do this in the abandoned [KEP-1664]
+  (which this KEP is a spiritual successor to), but if we want that,
+  it can be done later, independently of these changes. (These changes
+  were mostly desired in the context of non-cloud-provider clusters
+  anyway.)
+
+[KEP-1664]: https://github.com/kubernetes/enhancements/issues/1664
+
+## Proposal
+
+### Risks and Mitigations
+
+As the intention is to not change the user-visible behavior except in
+clusters where administrators explicitly make use of the new
+functionality, there should be no risk of breaking existing clusters,
+nor of surprising administrators by suddenly exposing node services on
+unexpected IPs.
+
+## Design Details
+
+### Current behavior
+
+Currently, when `--cloud-provider` is passed to kubelet, kubelet
+expects `--node-ip` to be either unset, or a single IP address. (If it
+is unset, that is equivalent to passing `--node-ip 0.0.0.0`, which
+means "autodetect an IPv4 address, or if there are no usable IPv4
+addresses, autodetect an IPv6 address".)
+
+If `--cloud-provider` and `--node-ip` are both specified (and
+`--node-ip` is not "`0.0.0.0`" or "`::`"), then kubelet will add an
+annotation to the node, `alpha.kubernetes.io/provided-node-ip`. Cloud
+providers expect this annotation to conform to the current expected
+`--node-ip` syntax (ie, a single value); if it does not, then they
+will log an error and, not remove the
+`node.cloudprovider.kubernetes.io/uninitialized` taint from the node,
+causing the node to remain unusable until kubelet is restarted with a
+valid (or absent) `--node-ip`.
+
+When `--cloud-provider` is not passed, the `--node-ip` value can also
+be a comma-separated pair of dual-stack IP addresses. However, unlike
+in the single-stack case, the IPs in the dual-stack case are not
+currently allowed to be "unspecified" IPs (ie `0.0.0.0` or `::`); you
+can only make a (non-cloud) node be dual-stack if you explicitly
+specify both IPs that you want it to use.
+
+### Changes to `--node-ip`
```

Comment:
Can you provide an example of what the flag would look like in the new version? Would I do something like `--node-ip 1.2.3.4 --node-ip xx::xx` to specify the multiple values?

Reply:
So customer now can use below two kind of options after this KEP either without cloud-provider or with cloud-provider?
1. --node-ip 1.2.3.4,xx::xx
2. --node-ip IPv4,IPv6

Reply:
And `--node-ip 1.2.3.4,IPv6` too. ("I want dual-stack IPv4-primary, and I want the IPv4 node IP to be 1.2.3.4, but I don't care what the IPv6 node IP is; just use the default".)

(As opposed to `--node-ip 1.2.3.4` which you can say now, but that means "use 1.2.3.4 as the IPv4 node IP, and be single-stack IPv4")

Reply:
a table will help a lot to visualize the matrix, we'll have to add it as a test later

|val1   |val2   | output   |  
|---|---|---|
|  1.2.3.4 |  2001:db8 |  ... |  
|  1.2.3.4 |  IPv6|  ... |  
|  0.0.0.0 |   :: |  ... |